kep/9-capacity-scheduling/README.md

```diff
@@ -320,6 +321,233 @@ preemptor’s priority as potential victims in a node.
+
+#### 1. Cross Node Preemption is not supported
+Current default Kubernetes scheduler does not support cross node preemption, which means that preemption process only happens in one node. In order to make capacity scheduler compatible with the default scheduler, current implementation also does not support cross node preemption.
+
+Because of the inability to do cross node preemption, it is possible that the scheduler fails to give the resources back to the original namespace.
+
+Below is a simple example. Suppossed that you have **exactly** 2 nodes in your cluster with the following configuration:
+##### Elastic Quota Configuration for namespace 1
+```
+apiVersion: v1
+kind: ElasticQuota
+metadata:
+  name: quota1
+  namespace: quota1
+spec:
+  max:
+    cpu: 2
+  min:
+    cpu: 0
+```
+##### Elastic Quota Configuration for namespace 2
+```
+apiVersion: v1
+kind: ElasticQuota
+metadata:
+  name: quota2
+  namespace: quota2
+spec:
+  max:
+    cpu: 2
+  min:
+    cpu: 2
+```
+##### Run Sample Deployment on namespace 1
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx
+  namespace: quota1
+  labels:
+    app: nginx
+spec:
+  replicas: 2
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app: nginx
+    spec:
+      schedulerName: capacityscheduler
+      containers:
+        - name: nginx
+          image: nginx
+          resources:
+            limits:
+              cpu: 1
+            `requests:`
+              cpu: 1
+```
+Suppossed that after running this deployment on namespace quota1, one replica is on Node 1 and another replica is on Node 2, which is highly the case.
+
+
+##### Run Sample Deployment on namespace 2
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx
+  namespace: quota2
+  labels:
+    app: nginx
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app: nginx
+    spec:
+      schedulerName: capacityscheduler
+      containers:
+        - name: nginx
+          image: nginx
+          resources:
+            limits:
+              cpu: 2
+            `requests:`
+              cpu: 2
+```
+
+Currently, namespace quota1 is overusing its min and borrowing the 2 unused cpu min from namespace quota2. Idealy, the two pods in namespace quota1 should get preempted and the newly deployed pod on namespace quota2 should be scheduled. However because preemption can only happen on one node, preempting one pod in namespace quota1 from only one node and running the newly deployed pod in namespace quota2 on that node will violate one of our rules. It will cause the total used cpu = 3 > 2 (sum of elastic quota min). Thus in this case, the preemption process fails no matter which node is selected for the preemption process.
```

Comment:
As you may have noticed, namespace quota1 overused its min quota, and has borrowed 2 cpus from quota2. Now here comes a pod from namespace quota2: ideally, the two pods in namespace quota1 should get preempted and the newly deployed pod on namespace quota2 should be scheduled. However, preemption is restricted to be attempted on one single node, so preempting either pod of namespace quota1 won't satisfy the incoming pod of quota2 - used cpus = 3 > 2 (sum of elastic quota min). Thus in this case, the preemption process fails no matter which node is selected.